lib/WUI/espif.cpp

```diff
@@ -0,0 +1,538 @@
+                mac_len = MAC_DATA_MAX_LEN;
+            }
+            mac_read = 0;
+            state = MACData;
+            log_debug(ESPIF, "Reading MAC len: %d", mac_len);
+            // No break, with zero len MAC data there might be no other input byte
+            // to trigger MAC processing.
+
+        case MACData:
+            while (c < end && mac_read < mac_len) {
+                log_debug(ESPIF, "Read MAC byte at %d: %02x", mac_read, *c);
+                mac_data[mac_read++] = *c++;
+            }
+            if (mac_read == mac_len) {
+                process_mac(mac_len, mac_data, netif);
+                state = Intron;
+            }
+            break;
+
+        case Packet:
+            rx_len_read = 0;
+            state = PacketLen;
+            break;
+
+        case PacketLen:
+            if (rx_len_read < sizeof(rx_len)) {
+                ((uint8_t *)&rx_len)[rx_len_read++] = *c++;
+            } else {
+                log_debug(ESPIF, "Reading packet size: %d", rx_len);
+#if ETH_PAD_SIZE
+                rx_len += ETH_PAD_SIZE; /* allow room for Ethernet padding */
+#endif
+                rx_buff = pbuf_alloc(PBUF_RAW, rx_len, PBUF_POOL);
+                if (rx_buff) {
+                    rx_buff_cur = rx_buff;
+                    rx_read = 0;
+                    state = PacketData;
+                } else {
+                    log_error(ESPIF, "Dropping packet due to out of RAM");
+                    state = Intron;
+                }
+            }
+            break;
+
+        case PacketData: {
+            // Copy input to current pbuf (until end of input or current pbuf)
+            const uint32_t to_read = std::min(rx_buff_cur->len - rx_read, (uint32_t)(end - c));
+            memcpy((uint8_t *)rx_buff_cur->payload + rx_read, c, to_read);
+            c += to_read;
+            rx_read += to_read;
+
+            // Switch to next pbuf
+            if (rx_read == rx_buff_cur->len) {
+                rx_buff_cur = rx_buff_cur->next;
+                rx_read = 0;
+            }
+
+            // Filled all pbufs in a packet (current set to next = NULL)
+            if (!rx_buff_cur) {
+                log_debug(ESPIF, "Read packet size: %d", rx_len);
+                if (netif->input(rx_buff, netif) != ERR_OK) {
+                    log_error(ESPIF, "ethernetif_input: IP input error");
+                    pbuf_free(rx_buff);
+                    state = Intron;
+                    break;
+                }
+                log_debug(ESPIF, "Input packet processed ok");
+                state = Intron;
+            }
+        } break;
+        }
+    }
+    log_debug(ESPIF, "Processed %d from UART", size);
+}
+
+/**
+ * @brief Send packet using ESPIF NIC
+ *
+ * @param netif Output NETIF handle
+ * @param p buffer (chain) to send
+ */
+static err_t low_level_output(struct netif *netif, struct pbuf *p) {
```

Comment:
Just to check:
• If the uart doesn't keep up, does this block?
• We are not supposed to free the pbuf, right? (I guess not, as it keeps working and by now I'd have run out of memory if we were).

Reply:
Yes this blocks (10ms per byte ~ 15s per 1500 byte frame). Occasionally this times out due to some weird dependency of RX and TX. Anyway "p" is not being used after this returns. As this takes no reference to "p" it is freed by LwIP once this returns. In future we bight increase reference counter on "p" and stockpile pbuf data for later transmit (by DMA?).

Reply:
Blocking is fine, we probably don't need to be too clever here. I just wanted to be sure we are not overwriting it faster than the other side can read and process and that things won't get lost. Hopefully.